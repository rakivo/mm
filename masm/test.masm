_start:
    push 10
    push 5
    push 2
    push 3
    push 8
    push 6
    push 9
    jmp start_operations

start_operations:
    dup 1       ; Duplicate the top stack element (9)
    dmp 1
    inc         ; Increment the top stack element by 1 (10)
    dmp 1
    swap 1       ; Swap the top two stack elements (6, 10)
    dmp 1
    pop         ; Remove the top stack element (6)
    dmp 1
    push 12
    push 4
    idiv        ; Divide 12 by 4 (result is 3)
    dmp 1
    push 15
    push 5
    imul        ; Multiply 15 by 5 (result is 75)
    dmp 1
    push 30
    push 3
    isub        ; Subtract 3 from 30 (result is 27)
    dmp 1
    push 40
    push 20
    iadd        ; Add 40 and 20 (result is 60)
    dmp 1

check_operations:
    cmp 27      ; Compare the top stack element with 27
    je equal_27
    jl less_than_27
    jg greater_than_27

equal_27:
    push 100
    dmp 1
    jmp next_section

less_than_27:
    push 50
    dmp 1
    jmp next_section

greater_than_27:
    push 200
    dmp 1

next_section:
    push 3.14   ; Push a floating point number
    push 2.71
    fadd        ; Add two floating point numbers
    dmp 1
    push 9.81
    push 1.62
    fmul        ; Multiply two floating point numbers
    dmp 1

    push 25
    push 5
    jmp divide_and_dump ; Jump to divide and dump section

further_operations:
    push 7
    push 8
    iadd
    dmp 1
    push 14
    push 2
    isub
    dmp 1
    push 10
    push 2
    imul
    dmp 1
    push 36
    push 6
    idiv
    dmp 1

loop_example:
    push 5

loop_start:
    dmp 1
    dec
    dup 1
    cmp 0
    jg loop_start

final_operations:
    ; Using a mathematical trick to get the number 69
    push 23
    push 3
    imul        ; Multiply 23 by 3 (result is 69)
    dmp 1

    ; Additional calculations to get the number 69
    push 46
    push 23
    iadd        ; Add 46 and 23 (result is 69)
    dmp 1

    push 138
    push 2
    idiv        ; Divide 138 by 2 (result is 69)
    dmp 1

cleanup:
    ; Clear the stack, leaving only the number 69
    pop
    pop
    dmp 1
    jmp halt

divide_and_dump:
    idiv
    dmp 1
    jmp further_operations

halt:
    halt
