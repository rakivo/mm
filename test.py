import os
import subprocess as sp

MASM_EXTENSION = '.masm'
MASM_BIN_EXTENSION = '.bin'
EXPECTED_EXTENSION = '.txt'

BUILD_DIR = 'build'
EXPECTED_DIR = 'expected'

MASM_DIR = 'masm'
DEMASM_DIR = 'demasm'

MASM = os.path.join(BUILD_DIR, MASM_DIR)
DEMASM = os.path.join(BUILD_DIR, DEMASM_DIR)

LOG_CMD   = '[CMD]'
LOG_INFO  = '[INFO]'
LOG_PANIC = '[PANIC]'
LOG_ERROR = '[ERROR]'

def cmd(args: list[str]) -> str:
    print(LOG_CMD, *args)
    result = sp.run(' '.join(args), text=True, shell=True, capture_output=True)

    if result.returncode != 0:
        print(LOG_PANIC, f'Process exited abnormally with code {result.returncode}')
        print(LOG_PANIC, result.stderr, end='')
        exit(1)

    return result.stdout

def build_mm():
    cmd(['make'])

def build_examples():
    cmd(['make examples'])

def test_examples():
    files = os.listdir(MASM_DIR)
    for f in files:
        name, ext = os.path.splitext(f)
        if ext != MASM_EXTENSION: continue

        masmf = os.path.join(MASM_DIR, name + MASM_EXTENSION)
        binf = os.path.join(BUILD_DIR, name + MASM_BIN_EXTENSION)

        cmd([MASM, masmf, '-o', binf])
        got_out = cmd([DEMASM, binf])

        with open(os.path.join(EXPECTED_DIR, name + EXPECTED_EXTENSION), 'r') as expected_out:
            got = ' '.join([line.strip() for line in got_out.split('\n')[:-2]])
            expected = ' '.join([line.strip() for line in expected_out.read().split('\n')])
            if got.strip() == expected.strip():
                print(f'{name}: OK')
            else:
                print(f'{name}: FAILED')
                print(f'Got: {got}, expected: {expected}')

if __name__ == '__main__':
    os.makedirs(EXPECTED_DIR, exist_ok=True)

    build_mm()
    build_examples()
    test_examples()
