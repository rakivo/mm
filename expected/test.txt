6
7
9
7
3
75
27
60
200
5.85
15.892200000000003
5
15
12
20
6
5
69
69
69
69
